bin/curriculum/delete_unit.rb

```diff
@@ -40,17 +40,29 @@ def parse_options
 
 require_relative '../../dashboard/config/environment'
 options.unit_names.each do |unit_nm|
+  puts "Processing unit [#{unit_nm}]..."
+
   unit_to_del = Unit.find_by(name: unit_nm)
   raise "Unit with name #{unit_nm} not found" if unit_to_del.nil?
-  raise "Only units in_development can be deleted as they wouldn't have usage and hence any DB references." unless unit_to_del.published_state == "in_development"
+  raise "Published units (state set as 'Stable') cannot be deleted as they could have active sections and progress." unless unit_to_del.get_published_state != "stable"
+
+  # Include additional warning if the script is not in one of the safe states where its not available for use externally.
+  if unit_to_del.get_published_state != "in_development"
+    print "Published state for Unit [#{unit_nm}] is not in_development, script might have usage. " \
+      "Please confirm with curriculum team and ensure comms has been sent before deletion. Continue? (Y/N)"
+    next unless gets.chomp.casecmp('y').zero?
+  end
```

Comment:
I think I see a couple of problems here. if the course is stable, but the unit is in development, we should not raise. also, if the course is in development, and the unit is NOT in development, then we shouldn't warn.

I think you have two main choices, one is to write a single, short condition that catches all bad cases and the other is to have different messages for the two cases.

if just one case, the original code is pretty good, and could just use a better comment. or you could do:
```
raise "..." unless unit.get_published_state != 'stable' || unit.published_state == 'in_development'
```

If you want to give just the right error message in all cases, it might be something like:
```
# this means "if the unit is a standalone unit"
if unit.is_course?
  raise "cannot delete standalone unit in stable state" unless unit.pubished_state != 'stable'
else
  raise "cannot delete unit in a stable course unless unit is hidden. to fix this..." if unit.get_published_state == 'stable' && unit.published_state != 'in_development'
``

Reply:
Good catch here, thank you! I like the more specific error, but shouldn't the condition to also allow for cases where course is not stable. For cases example, in case where course is in preview and unit within is in any state, we are probably ok to go ahead with deletion (unless there are section references)??.

Reply:
yes I may have missed a case. I think it depends on what you want the new behavior to be, if it is not the same as the original code in the repo.

Reply:
Actually, looking through this more, the get_published_state method returns the unit's published_state as long as it is not null. So

>  if the course is stable, but the unit is in development

This would not raise the exception in current implementation as `get_published_state` would return in_development

if the course is in development, and the unit is NOT in development, In this case as well, unless the unit is marked stable, there shouldn't be an exception.

I'm a bit confused here - happy to hop on a call if that is easier.

Reply:
> Actually, looking through this more, the get_published_state method returns the unit's published_state as long as it is not null

I see, I had missed this. good catch! if there is subtle behavior like this you are relying on, I think it will be helpful to explain in a comment (or if possible to do accurately, in the error message)

Reply:
Just including a comment for now as I'm also planning on following up with a change to not allow deletion of units that are still part of unit groups. That should simplify the script quiet a bit and can do more clean as part of that change.

Reply:
@vijayamanohararaj I am just remembering that the page that issues the delete command may not already have the unit id available on it. while it could be possible to provide it, I think it would also be acceptable to just go back to deleting by name. I would like to avoid the situation where the delete command accepts either name or id, though.

thank you for taking this on! sorry you ended up having to fix my mistake.

Reply:
sorry, it looks like i put this comment in the wrong PR. removing it here since we also talked about it in slack